bpo-37367: octal escapes applied inconsistently throughout the interpreter and lib

websurfer5 opened the pull request:
Add a range check for octal values in byte strings that raises ValueError if the value is more than 255 (0o377).

Tests for all 1-, 2-, and 3-digit octal values are provided (both valid and error cases).

<!--
Thanks for your contribution!
Please read this comment in its entirety. It's quite important.

# Pull Request title

It should be in the following format:

```
bpo-NNNN: Summary of the changes made
```

Where: bpo-NNNN refers to the issue number in the https://bugs.python.org.

Most PRs will require an issue number. Trivial changes, like fixing a typo, do not need an issue.

# Backport Pull Request title

If this is a backport PR (PR made against branches other than `master`),
please ensure that the PR title is in the following format:

```
[X.Y] <title from the original PR> (GH-NNNN)
```

Where: [X.Y] is the branch name, e.g. [3.6].

GH-NNNN refers to the PR number from `master`.

-->


<!-- issue-number: [bpo-37367](https://bugs.python.org/issue37367) -->
https://bugs.python.org/issue37367
<!-- /issue-number -->

websurfer5 committed:
bpo-37367: raise ValueError for byte strings containing octal
values greater than 255 (0o377)

jdemeyer commented on Lib/test/test_bytes.py:
```diff
@@ -14,6 +14,7 @@
 import tempfile
 import textwrap
 import unittest
+from ast import literal_eval as _literal_eval
```
Why the underscore?

websurfer5 replied:
> Why the underscore?

I copied a pattern I saw in other test programs.  It isn't necessary since literal_eval is not used anywhere else in the file.  I will remove it.

jdemeyer commented on Lib/test/test_bytes.py:
```diff
@@ -962,6 +963,43 @@ def test_translate(self):
         self.assertEqual(c, b'hllo')
 
+    def test_octal_values(self):
+        # bpo-37367: verify that octal values greater than 255
```
octal values -> octal escape sequences

websurfer5 replied:
I have made the requested changes.

jdemeyer commented on Misc/NEWS.d/next/Core and Builtins/2019-07-08-19-08-38.bpo-37367.SVtBhh.rst:
```diff
@@ -0,0 +1,2 @@
+Byte strings containing octal values greater than 255 will now raise
+ValueError.  Patch by (Jeffrey Kintscher).
```
Why is your name between parentheses?

jdemeyer replied:
Again, I would use the word "escape sequences" here.

websurfer5 replied:
> Why is your name between parentheses?

That's a typo.  I will remove it.

jdemeyer reviewed (changes requested):
Looks good, apart from a few nitpicks.

websurfer5 committed:
bpo-37367: changes per reviewer request:

1. use "octal escape sequences" instead of "octal values" in comments
2. remove underscore from _literal_eval

Also, use "0o1000-0o400" instead of "256" to make it clear to the reader
where the value came from.

taleinat commented on Misc/NEWS.d/next/Core and Builtins/2019-07-08-19-08-38.bpo-37367.SVtBhh.rst:
```diff
@@ -0,0 +1,2 @@
+Byte strings containing octal escape sequences with values greater than 255
+will now raise ValueError.  Patch by Jeffrey Kintscher.
```
It would be helpful to mention the previous behavior here.

taleinat commented on Lib/test/test_bytes.py:
```diff
@@ -14,6 +14,7 @@
 import tempfile
 import textwrap
 import unittest
+import ast
```
Please keep the imports sorted alphabetically.

serhiy-storchaka commented on Lib/test/test_bytes.py:
```diff
@@ -962,6 +963,45 @@ def test_translate(self):
         c = b.translate(None, delete=b'e')
         self.assertEqual(c, b'hllo')
 
+    def test_octal_values(self):
```
It is not correct place for such test. This file is for tests of bytes and bytearray methods, not parsing Python code.

`test_string_literals` is for parsing string and bytes literals. Add some tests there.

serhiy-storchaka commented on Lib/test/test_codecs.py:
```diff
@@ -1174,7 +1174,7 @@ def test_escape(self):
         check(br"[\101]", b"[A]")
         check(br"[\1010]", b"[A0]")
-        check(br"[\501]", b"[A]")
+        self.assertRaises(ValueError, decode, br"[\501]")
```
Move it to test_errors below.

serhiy-storchaka commented on Objects/bytesobject.c:
```diff
@@ -1185,6 +1185,11 @@ PyObject *_PyBytes_DecodeEscape(const char *s,
             }
+            if (c > 255) {
+                PyErr_Format(PyExc_ValueError,
+                             "octal value must be in range(0, 256)");
```
The error message should contain the position of the invalid octal escape sequence. It would be nice if look similar to error message for invalid hexadecimal escape sequence. For example:

"invalid octal escape sequence at position %zd"

serhiy-storchaka reviewed (changes requested):
Please add also a not in the What's New document. Since it is a breaking change it should be in section "Porting to Python 3.11".
